proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,328 @@
+
+func (rule *BinaryMemcacheRule) matchOpcode(oc byte) bool {
+	return bytes.Contains(rule.opCodes, []byte{oc})
+}
+
+func getMemcacheKey(packet []byte) (string, error) {
+	var keyLength uint16
+	keyBuf := bytes.NewReader(packet[2:4])
+	err := binary.Read(keyBuf, binary.BigEndian, &keyLength)
+	if err != nil {
+		return "", err
+	}
+	var extrasLength uint8
+	extrasLength = packet[4]
+
+	return string(packet[24+extrasLength : 24+uint16(extrasLength)+keyLength]), nil
+}
+
+// L7BinaryMemcacheRuleParser parses protobuf L7 rules to and array of BinaryMemcacheRule
+// May panic
+func L7BinaryMemcacheRuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		panic(fmt.Errorf("Can't get L7 rules"))
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br BinaryMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "opCode":
+				br.opCode = v
+				br.opCodes = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+			case "keyPrefix":
+				br.keyPrefix = v
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				panic(fmt.Errorf("Unsupported key: %s", k))
+			}
+		}
+		log.Infof("Parsed BinaryMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, strData []string, offset uint32) (proxylib.OpType, uint32) {
+	log.Infof("OnData with offset %d", offset)
+
+	//TODO: optimise this
+	data := ([]byte(strings.Join(strData, "")))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	headerMissing := headerSize - len(data)
+
+	if headerMissing > 0 {
+		log.Infof("Did not receive full header, need %d more bytes", headerMissing)
+		return proxylib.MORE, uint32(headerMissing)
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
```

Comment:
Since we are passing replies without parsing them you could move the reply pass logic right after getting to know the body length. There is no need wait to receive the full frame, you can return PASS with length as the full frame length even if the full frame has not been received yet.

Reply:
done